// @jazzer.js/instrumentor/dist/plugins/coverageVisitor.js
"use strict";
/*
 * Copyright 2026 Code Intelligence GmbH
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeCoverageVisitor = makeCoverageVisitor;
/**
 * Shared coverage instrumentation visitor.
 *
 * Both the CJS instrumentor (incrementCounter calls) and the ESM
 * instrumentor (direct array writes) inject counters at the same
 * AST locations.  This module captures that shared visitor shape
 * and lets each variant supply its own expression generator.
 */
const core_1 = require("@babel/core");
const types_1 = require("@babel/types");
/**
 * Build a Babel visitor that inserts a counter expression at every
 * branch point.  The caller decides what that expression looks like.
 */
function makeCoverageVisitor(makeCounterExpr) {
    function makeStmt() {
        return core_1.types.expressionStatement(makeCounterExpr());
    }
    function wrapWithCounter(stmt) {
        const counter = makeStmt();
        if ((0, types_1.isBlockStatement)(stmt)) {
            stmt.body.unshift(counter);
            return stmt;
        }
        return core_1.types.blockStatement([counter, stmt]);
    }
    return {
        Function(path) {
            if ((0, types_1.isBlockStatement)(path.node.body)) {
                path.node.body.body.unshift(makeStmt());
            }
        },
        IfStatement(path) {
            path.node.consequent = wrapWithCounter(path.node.consequent);
            if (path.node.alternate) {
                path.node.alternate = wrapWithCounter(path.node.alternate);
            }
            path.insertAfter(makeStmt());
        },
        SwitchStatement(path) {
            for (const caseClause of path.node.cases) {
                caseClause.consequent.unshift(makeStmt());
            }
            path.insertAfter(makeStmt());
        },
        Loop(path) {
            path.node.body = wrapWithCounter(path.node.body);
            path.insertAfter(makeStmt());
        },
        TryStatement(path) {
            if (path.node.handler) {
                path.node.handler.body.body.unshift(makeStmt());
            }
            path.insertAfter(makeStmt());
        },
        LogicalExpression(path) {
            if (!(0, types_1.isLogicalExpression)(path.node.left)) {
                path.node.left = core_1.types.sequenceExpression([
                    makeCounterExpr(),
                    path.node.left,
                ]);
            }
            if (!(0, types_1.isLogicalExpression)(path.node.right)) {
                path.node.right = core_1.types.sequenceExpression([
                    makeCounterExpr(),
                    path.node.right,
                ]);
            }
        },
        ConditionalExpression(path) {
            path.node.consequent = core_1.types.sequenceExpression([
                makeCounterExpr(),
                path.node.consequent,
            ]);
            path.node.alternate = core_1.types.sequenceExpression([
                makeCounterExpr(),
                path.node.alternate,
            ]);
            if ((0, types_1.isBlockStatement)(path.parent)) {
                path.insertAfter(makeStmt());
            }
        },
    };
}
