// @jazzer.js/bug-detectors/dist/internal/ssrf.js
"use strict";
/*
 * Copyright 2026 Code Intelligence GmbH
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SSRFConfig = void 0;
exports.hookTCPSocket = hookTCPSocket;
exports.hookUDPSocket = hookUDPSocket;
exports.hookUDPSocketSend = hookUDPSocketSend;
const assert_1 = __importDefault(require("assert"));
const core_1 = require("@jazzer.js/core");
const hooking_1 = require("@jazzer.js/hooking");
const configuration_1 = require("../configuration");
class SSRFConfig {
    constructor() {
        this._permittedTCPConnections = new Map();
        this._permittedUDPConnections = new Map();
    }
    /**
     * Add a permitted TCP connection to the config.
     * @param hostname - The hostname of the permitted connection.
     * @param port - The port of the permitted connection.
     */
    addPermittedTCPConnection(hostname, port) {
        return this.addPermittedConnection(this._permittedTCPConnections, hostname, port);
    }
    /**
     * Check if a TCP connection is permitted.
     * @param hostname - The hostname of the connection.
     * @param port - The port of the connection.
     */
    isPermittedTCPConnection(hostname, port) {
        // Even though ports should be numbers, node API allows strings as well.
        return this.isPermittedConnection(this._permittedTCPConnections, hostname, port);
    }
    /**
     * Add a permitted UDP connection to the config.
     * @param hostname - The hostname of the permitted connection.
     * @param port - The port of the permitted connection.
     */
    addPermittedUDPConnection(hostname, port) {
        return this.addPermittedConnection(this._permittedUDPConnections, hostname, port);
    }
    /**
     * Check if a UDP connection is permitted.
     * @param hostname - The hostname of the connection.
     * @param port - The port of the connection.
     */
    isPermittedUDPConnection(hostname, port) {
        return this.isPermittedConnection(this._permittedUDPConnections, hostname, port);
    }
    /**
     * Clear all TCP connections allowed so far.
     */
    clearPermittedTCPConnections() {
        this._permittedTCPConnections.clear();
        return this;
    }
    /**
     * Clear all UDP connections allowed so far.
     */
    clearPermittedUDPConnections() {
        this._permittedUDPConnections.clear();
        return this;
    }
    /**
     * Clear all TCP and UDP connections allowed so far.
     */
    clear() {
        this.clearPermittedTCPConnections();
        this.clearPermittedUDPConnections();
        return this;
    }
    addPermittedConnection(permittedConnections, hostname, port) {
        (0, assert_1.default)(this.isValidPort(port), "Port must be an integer between 0 and 65535");
        if (permittedConnections.has(hostname)) {
            const ports = permittedConnections.get(hostname);
            if (ports !== undefined && !ports.has(port)) {
                ports.add(port);
            }
        }
        else {
            const ports = new Set();
            ports.add(port);
            permittedConnections.set(hostname, ports);
        }
        return this;
    }
    isPermittedConnection(permittedConnections, hostname, port) {
        if (typeof port === "string") {
            try {
                port = parseInt(port);
            }
            catch (e) {
                return true;
            }
        }
        // The original function should handle invalid ports (usually by an error).
        // SSRF should not be reported in that case.
        if (port === undefined || !this.isValidPort(port))
            return true;
        return permittedConnections.get(hostname)?.has(port) ?? false;
    }
    isValidPort(port) {
        return (typeof port === "number" &&
            port >= 0 &&
            port <= 65535 &&
            Number.isInteger(port));
    }
}
exports.SSRFConfig = SSRFConfig;
const config = new SSRFConfig();
// Add this bug detector's config to the global Map configs.
configuration_1.bugDetectorConfigurations.set("ssrf", config);
(0, hooking_1.registerBeforeHook)("Socket.prototype.connect", "net", false, hookTCPSocket);
(0, hooking_1.registerBeforeHook)("Socket.prototype.connect", "dgram", false, hookUDPSocket);
(0, hooking_1.registerBeforeHook)("Socket.prototype.send", "dgram", false, hookUDPSocketSend);
function hookTCPSocket(_thisPtr, args, _id) {
    if (args.length === 1) {
        const firstArgument = args[0];
        if (firstArgument !== null && typeof firstArgument === "object") {
            const options = firstArgument;
            let host = options.host;
            let port = options.port;
            if (firstArgument instanceof Array) {
                if (host === undefined) {
                    host = firstArgument[0].host || "localhost";
                }
                if (port === undefined) {
                    port = firstArgument[0].port;
                }
            }
            detectSSRF(port, host, "Attempted connection via TCP");
        }
    }
    else if (args.length === 2) {
        // connect(options: SocketConnectOpts, connectionListener?: () => void): this;
        const firstArgument = args[0];
        if (typeof firstArgument === "object" && firstArgument !== null) {
            const options = firstArgument;
            detectSSRF(options.port, options.host, "Attempted connection via TCP");
        }
    }
    else if (args.length === 3) {
        // connect(port: number, host: string, connectionListener?: () => void): this;
        detectSSRF(args[0], args[1], "Attempted connection via TCP");
    }
}
function getConnectionState(socket) {
    for (const symbol of Object.getOwnPropertySymbols(socket)) {
        if (symbol.toString() === "Symbol(state symbol)") {
            return socket[symbol];
        }
    }
    throw new Error("Internal socket state is missing. This is a bug in the SSRF bug detector.");
}
function hookUDPSocket(thisPtr, args, _id) {
    const socket = thisPtr;
    if (socket?.type === undefined)
        return;
    // Type is only used to determine the default host.
    const defaultHost = socket.type === "udp4" ? "127.0.0.1" : "::1";
    const host = typeof args[1] === "string" ? args[1] : defaultHost;
    detectSSRF(args[0], host, "Attempted connection via UDP", true);
}
// Connection-less send using UDP sockets.
function hookUDPSocketSend(thisPtr, args, _id) {
    const state = getConnectionState(thisPtr);
    const offset = args[1];
    const length = args[2];
    let port = args[3];
    let host = args[4];
    const CONNECT_STATE_CONNECTED = 2;
    // This follows the same logic as the send function in the dgram module (dgram.js)
    // to obtain port and address
    if (state.connectState === CONNECT_STATE_CONNECTED) {
        if (typeof length === "number") {
            if (typeof port === "function") {
                port = undefined;
            }
        }
        if (port || host) {
            // When already connected, the original function will throw an error, so we're done here.
            return;
        }
    }
    else {
        if (!host && !(port && typeof port !== "function")) {
            port = offset;
            host = length;
        }
    }
    if (typeof host === "function") {
        host = undefined;
    }
    else if (host && typeof host !== "string") {
        return;
    }
    detectSSRF(port, host, "Attempted a connectionless send via UDP", true);
}
function detectSSRF(port, host, message, isUDP = false) {
    if (typeof port === "number" || typeof port === "string") {
        // The string check below is necessary for JavaScript, where the user can pass
        // hostnames of any type. In that case, we want the original function
        // to throw an error, and thus return from here.
        // Since we don't validate hostnames, strings that are not valid host names or valid
        // but adjusted internally (e.g. "123" will be internally converted to "0.0.0.123")
        // will be reported as SSRF findings. We treat every hostname that is a string as valid.
        if (typeof host === "string") {
            if (isUDP
                ? !config.isPermittedUDPConnection(host, port)
                : !config.isPermittedTCPConnection(host, port)) {
                (0, core_1.reportAndThrowFinding)(`Server Side Request Forgery (SSRF)\n` +
                    "     " +
                    message +
                    ` to '${host}' on port: '${port}'\n`);
            }
        }
    }
}
