// @jazzer.js/core/dist/dictionary.js
"use strict";
/*
 * Copyright 2026 Code Intelligence GmbH
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Dictionary = void 0;
exports.addDictionary = addDictionary;
exports.toEscapedString = toEscapedString;
exports.convertDictionaryEntry = convertDictionaryEntry;
exports.useDictionaryByParams = useDictionaryByParams;
const fs_1 = __importDefault(require("fs"));
const tmp_1 = __importDefault(require("tmp"));
const globals_1 = require("./globals");
/**
 * Dictionaries can be used to provide additional mutation suggestions to the
 * fuzzer.
 */
class Dictionary {
    constructor() {
        this._entries = [];
    }
    get entries() {
        return [...this._entries];
    }
    addEntries(dictionary) {
        this._entries.push(...dictionary);
    }
}
exports.Dictionary = Dictionary;
function getDictionary() {
    return (0, globals_1.getOrSetJazzerJsGlobal)("dictionary", new Dictionary());
}
function addDictionary(...dictionary) {
    getDictionary().addEntries(dictionary);
}
/**
 * Escapes a byte array to a string that can be used in a libFuzzer dictionary.
 * The format is a double-quoted string with escaped hex bytes.
 * @param byteArray
 * @returns The escaped string.
 *
 * Example:
 *    new Uint8Array([0,1,2,3]) will be converted to '"\\x00\\x01\\x02\\x03"'
 * Example:
 *    "Amazing" will be converted to '"\\x41\\x6d\\x61\\x7a\\x69\\x6e\\x67"'
 */
function toEscapedString(byteArray) {
    return ('"' +
        Array.from(byteArray, (byte) => {
            return "\\x" + byte.toString(16).padStart(2, "0");
        }).join("") +
        '"');
}
function convertDictionaryEntry(entry) {
    // Strings are converted to UTF-8 Uint8Arrays before escaping all according to libFuzzer dictionary format.
    // Background: Strings are encoded to UTF-8 here, which matches the way strings are produced from bytes by
    // the FuzzedDataProvider (by default), as well as the encoding used with sanitizer trace hooks for string
    // comparisons (see packages/fuzzer/trace.ts).
    return toEscapedString(typeof entry === "string" ? new TextEncoder().encode(entry) : entry);
}
function useDictionaryByParams(options, additionalDictionaryEntries = []) {
    const additionalDictionary = additionalDictionaryEntries.map(convertDictionaryEntry);
    const dictionary = getDictionary().entries.concat(additionalDictionary);
    // This diverges from the libFuzzer behavior, which allows only one dictionary (the last one).
    // We merge all dictionaries into one and pass that to libfuzzer.
    for (const option of options) {
        if (option.startsWith("-dict=")) {
            const dict = option.substring(6);
            // Preserve the filename in a comment before merging dictionary contents.
            dictionary.push(`\n# ${dict}:`);
            dictionary.push(fs_1.default.readFileSync(dict).toString());
        }
    }
    if (dictionary.length > 0) {
        // Add a comment to the top of the dictionary file.
        dictionary.unshift("# This file was automatically generated. Do not edit.");
        const content = dictionary.join("\n");
        // Use a temporary dictionary file to pass in the merged dictionaries.
        const dictFile = tmp_1.default.fileSync({
            mode: 0o700,
            prefix: "jazzer.js",
            postfix: "dict",
        });
        fs_1.default.writeFileSync(dictFile.name, content);
        fs_1.default.closeSync(dictFile.fd);
        return options.concat("-dict=" + dictFile.name);
    }
    return options;
}
