// @jazzer.js/core/dist/options.js
"use strict";
/*
 * Copyright 2026 Code Intelligence GmbH
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.OptionsManager = exports.OptionSource = exports.fromSnakeCaseWithPrefix = exports.fromSnakeCase = exports.fromCamelCase = exports.defaultJestOptions = exports.defaultCLIOptions = void 0;
exports.buildFuzzerOption = buildFuzzerOption;
exports.printOptions = printOptions;
exports.spawnsSubprocess = spawnsSubprocess;
exports.validateKeySource = validateKeySource;
const fs_1 = __importDefault(require("fs"));
const util = __importStar(require("util"));
const tmp = __importStar(require("tmp"));
const dictionary_1 = require("./dictionary");
const utils_1 = require("./utils");
// These options can be set from the Jest fuzz test.
const allowedFuzzTestOptions = [
    "dictionaryEntries",
    "fuzzerOptions",
    "sync",
    "timeout",
];
exports.defaultCLIOptions = Object.freeze({
    coverage: false,
    coverageDirectory: "coverage",
    coverageReporters: ["json", "text", "lcov", "clover"], // default Jest reporters
    customHooks: [],
    dictionaryEntries: [],
    disableBugDetectors: [],
    dryRun: false,
    excludes: ["node_modules"],
    expectedErrors: [],
    fuzzEntryPoint: "fuzz",
    fuzzerOptions: [],
    fuzzTarget: "",
    idSyncFile: "",
    includes: ["*"],
    mode: "fuzzing",
    sync: false,
    timeout: 5000, // default Jest timeout
    verbose: false,
});
exports.defaultJestOptions = Object.freeze({
    ...exports.defaultCLIOptions,
    mode: "regression",
});
const fromCamelCase = (key) => key;
exports.fromCamelCase = fromCamelCase;
const fromSnakeCase = (key) => {
    return (0, utils_1.replaceAll)(key.toLowerCase(), /(_[a-z0-9])/g, (group) => group.toUpperCase().replace("_", ""));
};
exports.fromSnakeCase = fromSnakeCase;
const fromSnakeCaseWithPrefix = (prefix) => {
    const prefixKey = prefix.toLowerCase() + "_";
    return (key) => {
        return key.toLowerCase().startsWith(prefixKey)
            ? (0, exports.fromSnakeCase)(key.substring(prefixKey.length))
            : key;
    };
};
exports.fromSnakeCaseWithPrefix = fromSnakeCaseWithPrefix;
// Source of an option is considered when merging options.
// Higher index means higher priority.
var OptionSource;
(function (OptionSource) {
    OptionSource[OptionSource["DefaultCLIOptions"] = 0] = "DefaultCLIOptions";
    OptionSource[OptionSource["DefaultJestOptions"] = 1] = "DefaultJestOptions";
    OptionSource[OptionSource["InternalJestTimeout"] = 2] = "InternalJestTimeout";
    OptionSource[OptionSource["ConfigurationFile"] = 3] = "ConfigurationFile";
    OptionSource[OptionSource["EnvironmentVariables"] = 4] = "EnvironmentVariables";
    OptionSource[OptionSource["CommandLineArguments"] = 5] = "CommandLineArguments";
    OptionSource[OptionSource["JestFuzzTestOptions"] = 6] = "JestFuzzTestOptions";
})(OptionSource || (exports.OptionSource = OptionSource = {}));
const defaultOptions = {
    [OptionSource.DefaultCLIOptions]: {
        name: "Default CLI options",
        transformKey: exports.fromCamelCase,
        failOnUnknown: true,
        parameters: exports.defaultCLIOptions,
    },
    [OptionSource.DefaultJestOptions]: {
        name: "Default Jest options",
        transformKey: exports.fromCamelCase,
        failOnUnknown: true,
        parameters: exports.defaultJestOptions,
    },
    [OptionSource.InternalJestTimeout]: {
        name: "Internal Jest timeout",
        transformKey: exports.fromCamelCase,
        failOnUnknown: true,
    },
    [OptionSource.ConfigurationFile]: {
        name: "Configuration file",
        transformKey: exports.fromCamelCase,
        failOnUnknown: true,
    },
    [OptionSource.EnvironmentVariables]: {
        name: "Environment variables",
        transformKey: (0, exports.fromSnakeCaseWithPrefix)("JAZZER"),
        failOnUnknown: false,
        parameters: process.env,
    },
    [OptionSource.CommandLineArguments]: {
        name: "Command line arguments",
        transformKey: exports.fromCamelCase,
        failOnUnknown: true,
    },
    [OptionSource.JestFuzzTestOptions]: {
        name: "Jest fuzz test options",
        transformKey: exports.fromCamelCase,
        failOnUnknown: true,
    },
};
class OptionsManager {
    /**
     * Manages merging of options from different sources.
     * WARNING: each fuzz test needs a copy (use the `clone()` function) of the OptionsManager, otherwise the fuzz tests will overwrite each other's options.
     * @param sourceOrOptions - build options given the `OptionSource`; or use provided options as is.
     */
    constructor(sourceOrOptions) {
        if (typeof sourceOrOptions === "number") {
            const source = sourceOrOptions;
            const initialOptions = defaultOptions[source].parameters;
            if (!initialOptions) {
                throw new Error(`Default options for ${source} do not exist. Consider adding them or use a different source.`);
            }
            this._options = OptionsManager.copyOptions(OptionsManager.attachSource(initialOptions, source));
            this.merge(process.env, OptionSource.EnvironmentVariables);
        }
        else if (typeof sourceOrOptions === "object") {
            // only used by clone()
            this._options = OptionsManager.copyOptions(sourceOrOptions);
        }
        else {
            throw new Error("Invalid argument");
        }
    }
    /**
     * Get the value of an option.
     * @param key
     */
    get(key) {
        return this._options[key].value;
    }
    /**
     * Get raw options without the source information.
     * @returns a copy of the options without source information
     */
    getOptions() {
        return OptionsManager.detachSource(this._options);
    }
    getOptionsWithSource() {
        return this._options;
    }
    /**
     * Merge new options from `input` given the `source` (aka priority). Same `source` options will result in an error---accumulate the options before writing.
     * `input` gets deep cloned to avoid reference keeping and unintended mutations.
     * @param input - new options to merge
     * @param source - priority of all the options in `input`
     */
    merge(input, source) {
        const transformKey = defaultOptions[source].transformKey;
        const errorOnUnknown = defaultOptions[source].failOnUnknown;
        let includes = undefined;
        let excludes = undefined;
        Object.keys(input).forEach((k) => {
            const transformedKey = transformKey(k);
            // Use hasOwnProperty to still support node v14.
            // eslint-disable-next-line no-prototype-builtins
            if (!exports.defaultCLIOptions.hasOwnProperty(transformedKey)) {
                if (errorOnUnknown) {
                    throw new Error(`Unknown Jazzer.js option '${k}'`);
                }
                return;
            }
            const key = transformedKey;
            if (!validateOptionPermissions(key, source, this._options)) {
                return;
            }
            const keyType = typeof exports.defaultCLIOptions[key];
            // No way to dynamically resolve the types here, use (implicit) any for now.
            // @ts-ignore
            let resultValue = input[k];
            // Try to parse strings as JSON values to support setting arrays and
            // objects via environment variables and command line arguments.
            if ([
                OptionSource.CommandLineArguments,
                OptionSource.EnvironmentVariables,
            ].includes(source) &&
                keyType !== "string" &&
                (typeof resultValue === "string" || resultValue instanceof String)) {
                try {
                    resultValue = JSON.parse(resultValue.toString());
                }
                catch (ignore) {
                    // Ignore parsing errors and continue with the string value.
                }
            }
            if (typeof resultValue !== keyType) {
                throw new Error(`Invalid type for Jazzer.js option '${key}', expected type '${keyType}', got '${typeof resultValue}'`);
            }
            // Deep copy the new value to avoid reference keeping and unintended mutations.
            resultValue = OptionsManager.copyOptionValue(resultValue);
            setProperty(this._options, key, { value: resultValue, source: source });
            if (key === "includes") {
                includes = resultValue;
            }
            else if (key === "excludes") {
                excludes = resultValue;
            }
        });
        // Includes and excludes must be set together.
        if (input && includes && !excludes) {
            this._options.excludes.value = [];
        }
        else if (input && excludes && !includes) {
            this._options.includes.value = [];
        }
        // Set verbose mode environment variable via option or node DEBUG environment variable.
        // Subsequent changes to the `verbose` option will be ignored.
        if (this.get("verbose") || process.env.DEBUG) {
            process.env.JAZZER_DEBUG = "1";
        }
        return this;
    }
    clone() {
        return new OptionsManager(this._options);
    }
    static copyOptions(newOptions) {
        const result = Object.create(null);
        Object.entries(newOptions).forEach(([k]) => {
            const key = k;
            const option = newOptions[key];
            const value = OptionsManager.copyOptionValue(option.value);
            const source = option.source;
            setProperty(result, key, {
                value,
                source,
            });
        });
        return result;
    }
    static copyOptionValue(input) {
        // simple types
        if (!input || typeof input !== "object") {
            return input;
        }
        if (Array.isArray(input)) {
            // (Uint8Array | Int8Array)[] - each sub-array gets copied
            if (input.some((element) => element instanceof Uint8Array || element instanceof Int8Array)) {
                return input.map((element) => {
                    if (element instanceof Uint8Array || element instanceof Int8Array) {
                        return element.slice();
                    }
                    return element;
                });
            }
            // string[] - the array can be copied directly
            return input.slice();
        }
        throw new Error("copyOptionValue: unsupported type: " + typeof input);
    }
    /**
     * Build options with source information attached.
     *
     * @param options
     * @returns a copy of the options with source information
     */
    static attachSource(options, source) {
        const result = Object.create(null);
        Object.entries(options).forEach(([k]) => {
            const key = k;
            setProperty(result, key, {
                value: options[key],
                source: source,
            });
        });
        return result;
    }
    /**
     * Remove source information from options.
     *
     * @param options
     * @returns a copy of the options without source information
     */
    static detachSource(options) {
        const result = Object.create(null);
        Object.entries(options).forEach(([k]) => {
            const key = k;
            const value = options[key]?.value;
            setProperty(result, key, value);
        });
        return result;
    }
}
exports.OptionsManager = OptionsManager;
function setProperty(obj, key, value) {
    obj[key] = value;
}
function buildFuzzerOption(options) {
    let params = [];
    params = optionDependentParams(options, params);
    params = forkedExecutionParams(params);
    params = (0, dictionary_1.useDictionaryByParams)(params, options.get("dictionaryEntries"));
    // libFuzzer has to ignore SIGINT and SIGTERM, as it interferes
    // with the Node.js signal handling.
    params = params.concat("-handle_int=0", "-handle_term=0", "-handle_segv=0");
    printOptions(options);
    logInfoAboutFuzzerOptions(params);
    return params;
}
function printOptions(options, infix = "") {
    if (process.env.JAZZER_DEBUG) {
        console.error(util.formatWithOptions(
        // Print everything in the options object.
        { maxArrayLength: null, depth: null, colors: false }, `DEBUG: [core] Jazzer.js options ${infix}: \n%O`, toOptionsWithPrintableSources(options)));
    }
}
function toOptionsWithPrintableSources(options) {
    const result = Object.create(null);
    const opts = options.getOptionsWithSource();
    Object.entries(opts).forEach(([k]) => {
        const key = k;
        const value = opts[key]?.value;
        const sourceIndex = opts[key]?.source;
        if (sourceIndex !== undefined) {
            const source = defaultOptions[sourceIndex].name;
            setProperty(result, key, { value, source });
        }
    });
    return result;
}
function logInfoAboutFuzzerOptions(fuzzerOptions) {
    fuzzerOptions.slice(1).forEach((element) => {
        if (element.length > 0 && element[0] != "-") {
            console.error("INFO: using inputs from:", element);
        }
    });
}
function optionDependentParams(options, params) {
    if (!options || !options.get("fuzzerOptions")) {
        return params;
    }
    let opts = options.get("fuzzerOptions");
    if (options.get("mode") === "regression") {
        // The last provided option takes precedence
        opts = opts.concat("-runs=0");
    }
    if (options.get("timeout") <= 0) {
        throw new Error("timeout must be > 0");
    }
    const inSeconds = Math.ceil(options.get("timeout") / 1000);
    opts = opts.concat(`-timeout=${inSeconds}`);
    return opts;
}
function forkedExecutionParams(params) {
    return [prepareLibFuzzerArg0(params), ...params];
}
function prepareLibFuzzerArg0(fuzzerOptions) {
    // When we run in a libFuzzer mode that spawns subprocesses, we create a wrapper script
    // that can be used as libFuzzer's argv[0]. In the fork mode, the main libFuzzer process
    // uses argv[0] to spawn further processes that perform the actual fuzzing.
    if (!spawnsSubprocess(fuzzerOptions)) {
        // Return a fake argv[0] to start the fuzzer if libFuzzer does not spawn new processes.
        return "unused_arg0_report_a_bug_if_you_see_this";
    }
    else {
        // Create a wrapper script and return its path.
        return createWrapperScript(fuzzerOptions);
    }
}
// These flags cause libFuzzer to spawn subprocesses.
const SUBPROCESS_FLAGS = ["fork", "jobs", "merge", "minimize_crash"];
function spawnsSubprocess(fuzzerOptions) {
    return fuzzerOptions.some((option) => SUBPROCESS_FLAGS.some((flag) => {
        const name = `-${flag}=`;
        return option.startsWith(name) && !option.startsWith("0", name.length);
    }));
}
function createWrapperScript(fuzzerOptions) {
    const jazzerArgs = process.argv.filter((arg) => arg !== "--" && fuzzerOptions.indexOf(arg) === -1);
    if (jazzerArgs.indexOf("--id_sync_file") === -1) {
        const idSyncFile = tmp.fileSync({
            mode: 0o600,
            prefix: "jazzer.js",
            postfix: "idSync",
        });
        jazzerArgs.push("--id_sync_file", idSyncFile.name);
        fs_1.default.closeSync(idSyncFile.fd);
    }
    const isWindows = process.platform === "win32";
    const scriptContent = `${isWindows ? "@echo off" : "#!/usr/bin/env sh"}
cd "${process.cwd()}"
${jazzerArgs.map((s) => '"' + s + '"').join(" ")} -- ${isWindows ? "%*" : "$@"}
`;
    const scriptTempFile = tmp.fileSync({
        mode: 0o700,
        prefix: "jazzer.js",
        postfix: "libfuzzer" + (isWindows ? ".bat" : ".sh"),
    });
    fs_1.default.writeFileSync(scriptTempFile.name, scriptContent);
    fs_1.default.closeSync(scriptTempFile.fd);
    return scriptTempFile.name;
}
// Check two things:
// 1) `dictionaryEntries` can only be set from "Jest fuzz test" source;
// 2) only few approved options can be set from "Jest fuzz test" source.
function validateKeySource(key, source) {
    const sourceName = defaultOptions[source].name;
    // Only "Jest fuzz test" is allowed to set `dictionaryEntries` option.
    if (key === "dictionaryEntries" &&
        source !== OptionSource.JestFuzzTestOptions) {
        const allowedSource = defaultOptions[OptionSource.JestFuzzTestOptions].name;
        throw new Error(`Tried setting option '${key}' from ${sourceName}, but this option is only available in ${allowedSource}`);
    }
    // Only selected options can be set from the Jest fuzz test
    if (source === OptionSource.JestFuzzTestOptions &&
        !allowedFuzzTestOptions.includes(key)) {
        throw new Error(`Option '${key}' is not available from "${sourceName}."`);
    }
}
// Check if the key can be set from the new source.
//
function validateOptionPermissions(key, source, options) {
    validateKeySource(key, source);
    // Overwriting options from the same source is not allowed---accumulate the options before writing.
    if (source === options[key].source) {
        throw new Error(`Option '${key}' already set from ${defaultOptions[source].name}`);
    }
    return source > options[key].source;
}
