// @jazzer.js/instrumentor/dist/instrument.js
"use strict";
/*
 * Copyright 2026 Code Intelligence GmbH
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Instrumentor = exports.MemorySyncIdStrategy = exports.FileSyncIdStrategy = exports.registerInstrumentationPlugin = exports.instrumentationGuard = void 0;
exports.registerInstrumentor = registerInstrumentor;
exports.registerEsmLoaderHooks = registerEsmLoaderHooks;
const path = __importStar(require("path"));
const url_1 = require("url");
const worker_threads_1 = require("worker_threads");
const core_1 = require("@babel/core");
const istanbul_lib_hook_1 = require("istanbul-lib-hook");
const hooking_1 = require("@jazzer.js/hooking");
const edgeIdStrategy_1 = require("./edgeIdStrategy");
const plugin_1 = require("./plugin");
const codeCoverage_1 = require("./plugins/codeCoverage");
const compareHooks_1 = require("./plugins/compareHooks");
const functionHooks_1 = require("./plugins/functionHooks");
const helpers_1 = require("./plugins/helpers");
const sourceCodeCoverage_1 = require("./plugins/sourceCodeCoverage");
const SourceMapRegistry_1 = require("./SourceMapRegistry");
var guard_1 = require("./guard");
Object.defineProperty(exports, "instrumentationGuard", { enumerable: true, get: function () { return guard_1.instrumentationGuard; } });
var plugin_2 = require("./plugin");
Object.defineProperty(exports, "registerInstrumentationPlugin", { enumerable: true, get: function () { return plugin_2.registerInstrumentationPlugin; } });
var edgeIdStrategy_2 = require("./edgeIdStrategy");
Object.defineProperty(exports, "FileSyncIdStrategy", { enumerable: true, get: function () { return edgeIdStrategy_2.FileSyncIdStrategy; } });
Object.defineProperty(exports, "MemorySyncIdStrategy", { enumerable: true, get: function () { return edgeIdStrategy_2.MemorySyncIdStrategy; } });
class Instrumentor {
    constructor(includes = [], excludes = [], customHooks = [], shouldCollectSourceCodeCoverage = false, isDryRun = false, idStrategy = new edgeIdStrategy_1.MemorySyncIdStrategy(), sourceMapRegistry = new SourceMapRegistry_1.SourceMapRegistry(), _seed = 3735928559) {
        this.includes = includes;
        this.excludes = excludes;
        this.customHooks = customHooks;
        this.shouldCollectSourceCodeCoverage = shouldCollectSourceCodeCoverage;
        this.isDryRun = isDryRun;
        this.idStrategy = idStrategy;
        this.sourceMapRegistry = sourceMapRegistry;
        this._seed = _seed;
        this.loaderPort = null;
        // This is our default case where we want to include everything and exclude the "node_modules" folder.
        if (includes.length === 0 && excludes.length === 0) {
            includes.push("*");
            excludes.push("node_modules");
        }
        this.includes = Instrumentor.cleanup(includes);
        this.excludes = Instrumentor.cleanup(excludes);
    }
    init() {
        (0, helpers_1.setSeed)(this._seed);
        if (this.includes.includes("jazzer.js")) {
            this.unloadInternalModules();
        }
        // Expose a registration function so ESM modules can feed their
        // source maps back to the main-thread registry.  The ESM loader
        // thread cannot access this registry directly, but the preamble
        // code it emits runs on the main thread during module evaluation
        // — before the module body, and therefore before any error could
        // need the map for stack-trace rewriting.
        const registry = this.sourceMapRegistry;
        globalThis.__jazzer_registerSourceMap = (filename, map) => registry.registerSourceMap(filename, map);
        return this.sourceMapRegistry.installSourceMapSupport();
    }
    instrument(code, filename, sourceMap) {
        // Extract inline source map from code string and use it as input source map
        // in further transformations.
        const inputSourceMap = sourceMap ?? (0, SourceMapRegistry_1.extractInlineSourceMap)(code);
        const transformations = [];
        const shouldInstrumentFile = this.shouldInstrumentForFuzzing(filename);
        if (shouldInstrumentFile) {
            transformations.push(...plugin_1.instrumentationPlugins.plugins, (0, codeCoverage_1.codeCoverage)(this.idStrategy), compareHooks_1.compareHooks);
        }
        if (hooking_1.hookManager.hasFunctionsToHook(filename)) {
            transformations.push((0, functionHooks_1.functionHooks)(filename));
        }
        if (this.shouldCollectCodeCoverage(filename)) {
            transformations.push((0, sourceCodeCoverage_1.sourceCodeCoverage)(filename, this.asInputSourceOption((0, SourceMapRegistry_1.toRawSourceMap)(inputSourceMap))));
        }
        if (shouldInstrumentFile) {
            this.idStrategy.startForSourceFile(filename);
        }
        let result = null;
        try {
            result = this.transform(filename, code, transformations, this.asInputSourceOption(inputSourceMap));
        }
        catch (e) {
            if (process.env.JAZZER_DEBUG) {
                const message = e instanceof Error ? e.message : e;
                console.error(`Instrumentation error in file ${filename}:\n  ${message}`);
            }
        }
        if (shouldInstrumentFile) {
            this.idStrategy.commitIdCount(filename);
        }
        return result;
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    asInputSourceOption(inputSourceMap) {
        // Empty input source maps mess up the coverage report.
        if (inputSourceMap) {
            return {
                inputSourceMap,
            };
        }
        return {};
    }
    transform(filename, code, plugins, options = {}) {
        if (plugins.length === 0) {
            return null;
        }
        const result = (0, core_1.transformSync)(code, {
            filename: filename,
            sourceFileName: filename,
            sourceMaps: true,
            plugins: plugins,
            ...options,
        });
        if (result?.map) {
            this.sourceMapRegistry.registerSourceMap(filename, result.map);
        }
        return result;
    }
    unloadInternalModules() {
        console.error("DEBUG: Unloading internal Jazzer.js modules for instrumentation...");
        [
            "@jazzer.js/bug-detectors",
            "@jazzer.js/core",
            "@jazzer.js/fuzzer",
            "@jazzer.js/hooking",
            "@jazzer.js/instrumentor",
            "@jazzer.js/jest-runner",
        ].forEach((module) => {
            delete require.cache[require.resolve(module)];
        });
    }
    shouldInstrumentForFuzzing(filepath) {
        return (!this.isDryRun &&
            Instrumentor.doesMatchFilters(filepath, this.includes, this.excludes));
    }
    get dryRun() {
        return this.isDryRun;
    }
    get includePatterns() {
        return this.includes;
    }
    get excludePatterns() {
        return this.excludes;
    }
    get coverageEnabled() {
        return this.shouldCollectSourceCodeCoverage;
    }
    get seed() {
        return this._seed;
    }
    /** Connect the main-thread side of the loader MessagePort. */
    setLoaderPort(port) {
        this.loaderPort = port;
    }
    /**
     * Send the current hook definitions to the ESM loader thread so it
     * can apply function-hook transformations.  Must be called after all
     * hooks are registered and finalized, but before user modules are
     * loaded.
     */
    sendHooksToLoader() {
        if (!this.loaderPort)
            return;
        const hooks = hooking_1.hookManager.hooks.map((hook, index) => ({
            id: index,
            type: hook.type,
            target: hook.target,
            pkg: hook.pkg,
            async: hook.async,
        }));
        this.loaderPort.postMessage({ hooks });
    }
    shouldCollectCodeCoverage(filepath) {
        return (this.shouldCollectSourceCodeCoverage &&
            (Instrumentor.doesMatchFilters(filepath, this.includes, this.excludes) ||
                Instrumentor.doesMatchFilters(filepath, this.customHooks, ["nothing"])));
    }
    static doesMatchFilters(filepath, includes, excludes) {
        const included = includes.find((include) => filepath.includes(include)) !== undefined;
        const excluded = excludes.find((exclude) => filepath.includes(exclude)) !== undefined;
        return included && !excluded;
    }
    static cleanup(settings) {
        return settings
            .filter((setting) => setting)
            .map((setting) => (setting === "*" ? "" : setting)); // empty string matches every file
    }
}
exports.Instrumentor = Instrumentor;
function registerInstrumentor(instrumentor) {
    instrumentor.init();
    (0, istanbul_lib_hook_1.hookRequire)(() => true, (code, opts) => {
        return instrumentor.instrument(code, opts.filename)?.code || code;
    }, { extensions: [".js", ".mjs", ".cjs", ".ts", ".mts", ".cts"] });
}
/**
 * Register an ESM loader hook so that import() and static imports are
 * instrumented too.  Only needed when modules are loaded via Node
 * directly (CLI path).  The jest-runner uses Jest's own transformer
 * pipeline instead and does not need the ESM loader thread.
 */
function registerEsmLoaderHooks(instrumentor) {
    registerEsmHooks(instrumentor);
}
function nodeVersionAtLeast(major, minor) {
    const [maj, min] = process.versions.node.split(".").map(Number);
    return maj > major || (maj === major && min >= minor);
}
/**
 * On Node.js >= 20.6 register an ESM loader hook so that
 * import() and static imports are instrumented too.
 *
 * On Node >= 20.11 (where module.register supports transferList) we
 * also establish a MessagePort to the loader thread.  This lets us
 * send function-hook definitions after bug detectors are loaded —
 * well before user modules are imported.
 */
function registerEsmHooks(instrumentor) {
    if (instrumentor.dryRun) {
        return;
    }
    // module.register() is available since Node 20.6.
    if (!nodeVersionAtLeast(20, 6)) {
        console.error("WARN: [jazzer.js] ES module instrumentation requires Node.js >= 20.6" +
            ` (current: ${process.versions.node}).` +
            " ES modules (import/import()) will not be instrumented.");
        return;
    }
    // transferList for MessagePort (needed to send function-hook
    // definitions to the loader thread) requires Node 20.11.  Without
    // it ESM still gets coverage and compare-hooks, but bug detectors
    // will not fire on ES module code.
    const canTransferPort = nodeVersionAtLeast(20, 11);
    if (!canTransferPort) {
        console.error("WARN: [jazzer.js] ES module function hooks require Node.js >= 20.11" +
            ` (current: ${process.versions.node}).` +
            " Bug detectors will not apply to ES modules.");
    }
    try {
        const { register } = require("node:module");
        const loaderUrl = (0, url_1.pathToFileURL)(path.join(__dirname, "esm-loader.mjs")).href;
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const data = {
            includes: instrumentor.includePatterns,
            excludes: instrumentor.excludePatterns,
            coverage: instrumentor.coverageEnabled,
            seed: instrumentor.seed,
        };
        const options = { parentURL: (0, url_1.pathToFileURL)(__filename).href, data };
        if (canTransferPort) {
            const { port1, port2 } = new worker_threads_1.MessageChannel();
            data.port = port2;
            options.transferList = [port2];
            instrumentor.setLoaderPort(port1);
        }
        register(loaderUrl, options);
    }
    catch (e) {
        const detail = e instanceof Error ? ` (${e.message})` : "";
        console.error("WARN: [jazzer.js] Failed to register ESM loader hooks." +
            " ES modules will not be instrumented." +
            detail);
        if (process.env.JAZZER_DEBUG && e instanceof Error) {
            console.error(e.stack);
        }
    }
}
