// @jazzer.js/hooking/dist/manager.js
"use strict";
/*
 * Copyright 2026 Code Intelligence GmbH
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.hookManager = exports.HookManager = exports.MatchingHooksResult = void 0;
exports.registerBeforeHook = registerBeforeHook;
exports.registerReplaceHook = registerReplaceHook;
exports.registerAfterHook = registerAfterHook;
exports.getFunction = getFunction;
exports.setFunction = setFunction;
exports.hookBuiltInFunction = hookBuiltInFunction;
exports.callSiteId = callSiteId;
const module_1 = require("module");
const vm = __importStar(require("vm"));
const hook_1 = require("./hook");
const tracker_1 = require("./tracker");
class MatchingHooksResult {
    constructor() {
        this._beforeHooks = [];
        this._replaceHooks = [];
        this._afterHooks = [];
    }
    get hooks() {
        return this._beforeHooks.concat(this._afterHooks, this._replaceHooks);
    }
    hasHooks() {
        return (this.hasBeforeHooks() || this.hasReplaceHooks() || this.hasAfterHooks());
    }
    get beforeHooks() {
        return this._beforeHooks;
    }
    hasBeforeHooks() {
        return this._beforeHooks.length !== 0;
    }
    get replaceHooks() {
        return this._replaceHooks;
    }
    hasReplaceHooks() {
        return this._replaceHooks.length !== 0;
    }
    get afterHooks() {
        return this._afterHooks;
    }
    hasAfterHooks() {
        return this._afterHooks.length !== 0;
    }
    addHook(h) {
        switch (h.type) {
            case hook_1.HookType.Before:
                this._beforeHooks.push(h);
                break;
            case hook_1.HookType.Replace:
                this._replaceHooks.push(h);
                break;
            case hook_1.HookType.After:
                this._afterHooks.push(h);
                break;
        }
    }
    verify() {
        if (this._replaceHooks.length > 1) {
            throw new Error(`For a given target function, one REPLACE hook can be configured. Found: ${this._replaceHooks.length}`);
        }
        if (this.hasReplaceHooks() &&
            (this.hasBeforeHooks() || this.hasAfterHooks())) {
            throw new Error(`For a given target function, REPLACE hooks cannot be mixed up with BEFORE/AFTER hooks. Found ${this._replaceHooks.length} REPLACE hooks and ${this._beforeHooks.length + this._afterHooks.length} BEFORE/AFTER hooks`);
        }
        if (this.hasAfterHooks()) {
            if (!this._afterHooks.every((h) => h.async) &&
                !this._afterHooks.every((h) => !h.async)) {
                throw new Error("For a given target function, AFTER hooks have to be either all sync or all async.");
            }
        }
    }
}
exports.MatchingHooksResult = MatchingHooksResult;
class HookManager {
    constructor() {
        this._hooks = [];
    }
    /**
     * Finalizes the registration of new hooks and performs necessary
     * initialization steps for the hooks to work. This method must be called
     * after all hooks have been registered.
     */
    async finalizeHooks(vmContext) {
        // Built-in functions cannot be hooked by the instrumentor, so that is
        // explicitly done here instead.
        // Loading build-in modules is asynchronous, so we need to wait, which
        // is not possible in the instrumentor.
        for (const builtinModule of module_1.builtinModules) {
            const matchedHooks = this._hooks.filter((hook) => builtinModule.includes(hook.pkg) && hook.pkg !== "");
            for (const hook of matchedHooks) {
                try {
                    await hookBuiltInFunction(hook);
                }
                catch (e) {
                    if (process.env.JAZZER_DEBUG) {
                        console.error("DEBUG: [hooking] Could not hook built-in function: " +
                            hook.pkg +
                            " : " +
                            hook.target);
                        console.error(e);
                    }
                }
            }
        }
        // Hook global functions (functions without module names)
        const globalHooks = this._hooks.filter((hook) => hook.pkg === "");
        for (const hook of globalHooks) {
            try {
                hookGlobalFunction(hook, vmContext);
            }
            catch (e) {
                if (process.env.JAZZER_DEBUG) {
                    console.error("DEBUG: [hooking] Could not hook the global function: " +
                        hook.target);
                    console.error(e);
                }
            }
        }
    }
    registerHook(hookType, target, pkg, async, hookFn) {
        const hook = new hook_1.Hook(hookType, target, pkg, async, hookFn);
        this._hooks.push(hook);
        return hook;
    }
    get hooks() {
        return this._hooks;
    }
    clearHooks() {
        this._hooks = [];
    }
    hookIndex(hook) {
        return this._hooks.indexOf(hook);
    }
    matchingHooks(target, filepath) {
        const matches = this._hooks
            .filter((hook) => hook.match(filepath, target))
            .reduce((matches, hook) => {
            matches.addHook(hook);
            return matches;
        }, new MatchingHooksResult());
        matches.verify();
        return matches;
    }
    hasFunctionsToHook(filepath) {
        return (this._hooks.find((hook) => filepath.includes(hook.pkg)) !== undefined);
    }
    callHook(id, thisPtr, params, resultOrOriginalFunction) {
        const hook = this._hooks[id];
        switch (hook.type) {
            case hook_1.HookType.Before:
                hook.hookFunction(thisPtr, params, callSiteId());
                break;
            case hook_1.HookType.Replace:
                return hook.hookFunction(thisPtr, params, callSiteId(), 
                // eslint-disable-next-line @typescript-eslint/no-unsafe-function-type
                resultOrOriginalFunction);
            case hook_1.HookType.After:
                hook.hookFunction(thisPtr, params, callSiteId(), resultOrOriginalFunction);
        }
    }
}
exports.HookManager = HookManager;
exports.hookManager = new HookManager();
function registerBeforeHook(target, pkg, async, hookFn) {
    exports.hookManager.registerHook(hook_1.HookType.Before, target, pkg, async, hookFn);
}
function registerReplaceHook(target, pkg, async, hookFn) {
    exports.hookManager.registerHook(hook_1.HookType.Replace, target, pkg, async, hookFn);
}
function registerAfterHook(target, pkg, async, hookFn) {
    exports.hookManager.registerHook(hook_1.HookType.After, target, pkg, async, hookFn);
}
function getFunction(module, propertyAccessors) {
    let current = module;
    for (const propertyAccessor of propertyAccessors) {
        try {
            // @ts-ignore
            current = current[propertyAccessor];
        }
        catch (e) {
            return undefined;
        }
    }
    return current;
}
function setFunction(module, propertyAccessors, newFunction) {
    if (!(newFunction instanceof Function || typeof newFunction === "function")) {
        if (process.env.JAZZER_DEBUG) {
            console.error("DEBUG: [hooking] Could not hook built-in function: " +
                propertyAccessors.join(".") +
                "\n" +
                "   provided newFunction is not a function");
        }
        return;
    }
    let current = module;
    for (const propertyAccessor of propertyAccessors.slice(0, -1)) {
        try {
            // @ts-ignore
            current = current[propertyAccessor];
        }
        catch (e) {
            if (process.env.JAZZER_DEBUG) {
                console.error("DEBUG: [hooking] Could not hook built-in function: " +
                    propertyAccessors.join("."));
            }
            return;
        }
    }
    // @ts-ignore
    current[propertyAccessors[propertyAccessors.length - 1]] = newFunction;
}
/**
 * Replaces a built-in function with a custom implementation while preserving
 * the original function for potential use within the replacement function.
 */
async function hookBuiltInFunction(hook) {
    if (hook.registered)
        return;
    hook.registered = true;
    const { default: module } = await import(hook.pkg);
    const targetPropertyAccessors = hook.target.split(".");
    const originalFn = getFunction(module, targetPropertyAccessors);
    hookFunction(module, hook, originalFn, targetPropertyAccessors);
}
function hookGlobalFunction(hook, context) {
    const originalFn = vm.isContext(context)
        ? vm.runInContext(hook.target, context)
        : vm.runInThisContext(hook.target);
    if (originalFn) {
        hookFunction(context, hook, originalFn);
        const newFunction = getFunction(context, hook.target.split("."));
        // assign prototype of the original function to the prototype of the new function
        if (newFunction instanceof Function || typeof newFunction === "function") {
            newFunction.prototype = originalFn.prototype;
        }
    }
}
function hookFunction(
// eslint-disable-next-line @typescript-eslint/no-explicit-any
obj, hook, originalFn, targetPropertyAccessors = hook.target.split(".")) {
    if (!(originalFn instanceof Function || typeof originalFn === "function")) {
        if (process.env.JAZZER_DEBUG) {
            console.error("DEBUG: [hooking] Could not hook built-in function: " +
                hook.pkg +
                " : " +
                hook.target);
        }
        return;
    }
    const id = callSiteId(exports.hookManager.hookIndex(hook), hook.pkg, hook.target);
    if (hook.type == hook_1.HookType.Before) {
        setFunction(obj, targetPropertyAccessors, function (...args) {
            // @ts-ignore
            hook.hookFunction(this, args, id);
            // @ts-ignore
            return originalFn.apply(this, args);
        });
    }
    else if (hook.type == hook_1.HookType.Replace) {
        setFunction(obj, targetPropertyAccessors, function (...args) {
            // @ts-ignore
            return hook.hookFunction(this, args, id, originalFn);
        });
    }
    else if (hook.type == hook_1.HookType.After) {
        setFunction(obj, targetPropertyAccessors, function (...args) {
            // @ts-ignore
            const result = originalFn.apply(this, args);
            // @ts-ignore
            return hook.hookFunction(this, args, id, result);
        });
    }
    else {
        throw new Error(`Unknown hook type ${hook.type}`);
    }
    (0, tracker_1.logHooks)([hook]);
    tracker_1.hookTracker.addApplied(hook.pkg, hook.target);
}
/**
 * Returns a unique id for the call site of the function that called this function.
 * @param additionalArguments additional arguments to be included in the hash
 */
function callSiteId(...additionalArguments) {
    const stackTrace = additionalArguments?.join(",") + new Error().stack;
    if (!stackTrace || stackTrace.length === 0) {
        return 0;
    }
    let hash = 0, i, chr;
    for (i = 0; i < stackTrace.length; i++) {
        chr = stackTrace.charCodeAt(i);
        hash = (hash << 5) - hash + chr;
        hash |= 0; // Convert to 32bit integer
    }
    return hash;
}
